src/registrar/views/application.py

```diff
@@ -0,0 +1,421 @@
+    between pages of the application and provides common methods for
+    processing form data.
+
+    Views for each step should inherit from this base class.
+
+    Any method not marked as internal can be overridden in a subclass,
+    although not without consulting the base implementation, first.
+    """
+
+    # uniquely namespace the wizard in urls.py
+    # (this is not seen _in_ urls, only for Django's internal naming)
+    # NB: this is included here for reference. Do not change it without
+    # also changing the many places it is hardcoded in the HTML templates
+    URL_NAMESPACE = "application"
+    # name for accessing /application/<id>/edit
+    EDIT_URL_NAME = "edit-application"
+
+    # We need to pass our human-readable step titles as context to the templates.
+    TITLES = {
+        Step.ORGANIZATION_TYPE: _("Type of organization"),
+        Step.ORGANIZATION_FEDERAL: _("Type of organization — Federal"),
+        Step.ORGANIZATION_ELECTION: _("Type of organization — Election board"),
+        Step.ORGANIZATION_CONTACT: _("Organization name and mailing address"),
+        Step.AUTHORIZING_OFFICIAL: _("Authorizing official"),
+        Step.CURRENT_SITES: _("Organization website"),
+        Step.DOTGOV_DOMAIN: _(".gov domain"),
+        Step.PURPOSE: _("Purpose of your domain"),
+        Step.YOUR_CONTACT: _("Your contact information"),
+        Step.OTHER_CONTACTS: _("Other contacts for your domain"),
+        Step.SECURITY_EMAIL: _("Security email for public use"),
+        Step.ANYTHING_ELSE: _("Anything else we should know?"),
+        Step.REQUIREMENTS: _(
+            "Requirements for registration and operation of .gov domains"
+        ),
+        Step.REVIEW: _("Review and submit your domain request"),
+    }
+
+    # We can use a dictionary with step names and callables that return booleans
+    # to show or hide particular steps based on the state of the process.
+    WIZARD_CONDITIONS = {
+        Step.ORGANIZATION_FEDERAL: lambda w: w.from_model(
+            "show_organization_federal", False
+        ),
+        Step.ORGANIZATION_ELECTION: lambda w: w.from_model(
+            "show_organization_election", False
+        ),
+    }
+
+    def __init__(self):
+        super().__init__()
+        self.steps = StepsHelper(self)
+        self._application = None  # for caching
+
+    @property
+    def prefix(self):
+        """Namespace the wizard to avoid clashes in session variable names."""
+        # this is a string literal but can be made dynamic if we'd like
+        # users to have multiple applications open for editing simultaneously
+        return "wizard_application"
+
+    @property
+    def application(self) -> DomainApplication:
+        """
+        Attempt to match the current wizard with a DomainApplication.
+
+        Will create an application if none exists.
+        """
+        if self._application:
+            return self._application
+
+        if self.has_pk():
```

Comment:
**suggestion (non-blocking):** For locality of reference, could we move the definition of the `has_pk` method above this property? It mixes an ordinary method in with a section full of properties, but it felt slightly strange to have to look forward to find out what that method does.

Reply:
Moved `has_pk`; that was a good suggestion.